docs/benchmark/ftlib_benchmark.md

```diff
@@ -0,0 +1,171 @@
+those tensors have 2,236,682 trainable parameters.
+
+| Workers   |  computation/communication  |  Speed   |  Speedup Ratio |
+| -- | --------------------------- | -------- | ------ |
+| 1  |   -  | 29 images/s  |  1  |
+| 2  | 10: 3 | 44.7 images/s | 1.54 |
+| 3  | 10: 6 | 57.2 images/s  | 1.97 |
+
+## On-premise CPU Cluster Managed by a Tailored Kubernetes System
+
+There are more resource on an on-premise CPU cluster, so we can turn up the
+worker resource to do experiments. And, we also use CIFAR-10 dataset and
+assign more CPUs and memory for the worker. The experiment parameters are:
+
+| Experiment parameter | Value |
+| -- | --- |
+| Batch size | 64 |
+| Batches per task | 50 |
+| Images per task | 3220 |
+| Dataset | CIFAR-10 |
+| Input shape | (32, 32, 3) |
+| Output classes| 10 |
+| Worker resource| cpu=4,memory=8192Mi,ephemeral-storage=1024Mi|
+
+### ResNet50 with CIFAR-10 on an On-premise CPU Cluster
+
+| Workers   | communication   |  Speed   |  Speedup Ratio |
+| -- | --------------------------- | -------- | ------ |
+| 1  |   0%  | 26.7 images/s  |  1  |
+| 2  | 18% | 41 images/s | 1.57 |
+| 4  | 25% | 68.4 images/s  | 2.56 |
+| 8  | 32% | 123 images/s | 4.61 |
+
+### MobileNetV2 with CIFAR-10 on an On-premise CPU Cluster
+
+| Workers   | communication   |  Speed   |  Speedup Ratio |
+| -- | --------------------------- | -------- | ------ |
+| 1  |   0%  | 353.6 images/s  |  1  |
+| 2  | 24% | 503 images/s | 1.42 |
+| 4  | 44.7% | 680 images/s  | 1.92 |
+| 8  | 66.7% | 648 images/s  | 1.83 |
+
+For the above experiments of CPU, we find that the speed-up ratio is better if
+the model is more complex. Because the computation on CPUs is
+very slow. The computation is the bottleneck on the CPU cluster. So, the
+the speed-up of ResNet50 is better than MobileNetV2.
+
+## On-premise GPU Cluster Managed by a Tailored Kubernetes System
+
+On the on-premise GPU cluster, we can launch pods with single P100 GPU.
+As we known, GPUs are faster for computing than CPUs and the size of
+CIFAR-10 is small to test the performance with GPUs. We use ImageNet with
+input shape (256,256,3) to train Resnet50, MobileNetV2, and an iamge
+compression model. The last model has less parameters tensors.
+
+Experiment parameters are:
+
+| Experiment parameter | Value    |
+| -- | --- |
+| Batch size | 64 |
+| Batches per task | 16 |
+| Images per task | 1024 |
+| Dataset | ImageNet |
+| Image shape | (256, 256, 3) |
+| Output classes| 100 |
+| Worker resource| cpu=8,gpu=1,memory=16000Mi,ephemeral-storage=1024Mi|
+
+### ResNet50 with ImageNet on an On-premise GPU Cluster
+
+There are also 214 trainable tensors in ResNet50 for ImageNet. But the
+model trainable parameters is 23,739,492 because there are 100 classes.
+
+| Workers   | speed | total task time  | allreduce time| tensor.numpy() time| apply_gradients |
+| --------- | ----- | --------------- | -------- | ------ | ---------- |
+| 1 (local) | 168 images/s| 6.1s  |  - | - |  4.16s |
```

Comment:
Is `tensor.numpy()` time is a part of allreduce time?

Totoal task time = forward/backward time + allreduce time + apply gradients time

Reply:
Yes, it is.